src/blocks/speedtest.rs

```diff
@@ -203,22 +159,16 @@ impl Block for SpeedTest {
                 let ping = vals[0] as f64 / 1_000.0;
                 let down = vals[1] as f64 * 1_000_000.0;
                 let up = vals[2] as f64 * 1_000_000.0;
-                self.text[0].set_text(format_number(ping, self.config.speed_digits, "", "s"));
-                self.text[1].set_text(format_number(
-                    down,
-                    self.config.speed_digits,
-                    &self.config.speed_min_unit.to_string(),
-                    if self.config.bytes { "B/s" } else { "b/s" },
-                ));
-                self.text[2].set_text(format_number(
-                    up,
-                    self.config.speed_digits,
-                    &self.config.speed_min_unit.to_string(),
-                    if self.config.bytes { "B/s" } else { "b/s" },
-                ));
+
+                let values = map!(
+                    "ping" => Value::from_float(ping).seconds().icon(self.ping_icon.clone()),
+                    "speed_down" => Value::from_float(down).bits_per_second().icon(self.down_icon.clone()),
+                    "speed_up" => Value::from_float(up).bits_per_second().icon(self.up_icon.clone()),
+                );
+                self.text.set_text(self.format.render(&values)?);
 
                 // ping is in seconds
-                self.text[0].set_state(match (ping * 1000.) as i32 {
+                self.text.set_state(match (ping * 1000.) as i32 {
```

Comment:
@GladOSkar You mentioned in another comment:

> speedtest now colors the entire block based on ping instead of just 1 widget. I assume this is not fixable, but should be accounted for with some saner defaults

The reason is because speedtest was changed to use a format string by combining the two text widgets into one. The previous functionality should be able to kept if we keep the two widgets, and instead have one format string for each widget (Similar to how the NetworkManager block does it: https://github.com/greshake/i3status-rust/blob/master/blocks.md#networkmanager)

@MaxVerevkin, @GladOSkar Reckon we should go with that?

One disadvantage would be that the order of the widgets would be unconfigurable (as it is now), but that could be fixed by adding another config option for the widget order. This would allow us to have the best of both worlds.

Reply:
So your idea is basically to split `format` into
``` toml
format_ping = "{ping}"
format_speed = "{speed_down} {speed_up}"
```
I think It's good enough.

Reply:
Or maybe I have a different solution: we can just use pango's color attributes. This way, `Value` can be extended to have a background and foreground color. I even prefer this solution as it is more generic.

Reply:
That would be very neat of course but i doubt that'll render the color across the entire bar height, and instead only color the text height. But it's worth a shot.

Reply:
@GladOSkar Can you test 9f50e09? It works for me just fine.

Reply:
Perrsonally, i always disabled ping coloring entirely because i found it distracting and also not very meaningful. But that's probably not viable for everyone

Reply:
Nope, on `i3` it's as i feared:
![image](https://user-images.githubusercontent.com/9497832/111357466-a281f700-8689-11eb-9165-ff4160169b8c.png)

Reply:
That's what I have on `sway`
![image](https://user-images.githubusercontent.com/34583604/111357838-58e9da00-8692-11eb-94e0-9085029e5f16.png)

Reply:
> the neighboring placeholder is also not colored.

Should it?

Reply:
> That's what I have on sway

Yeah we had similar issues before. Sway renders quite differently from i3, since the latter has sort of 2 border rows that aren't part of the pango render area.

Reply:
> Should it be?

Yes. See `master`:
![image](https://user-images.githubusercontent.com/9497832/111360000-8469c600-868c-11eb-897f-8731e9021fce.png)

Reply:
Oh, you mean the separator. Well, this *is* unfixable without splitting the block into two widgets.

Reply:
Oh yeah sorry i confused those for a sec there

Reply:
But yeah the "widget" paradigm and the "format string" paradigm seem to be somewhat opposing at the moment. Do you think there is a way we could specify widgets within the format string? That seems a bit overkill. So probably multiple format strings are the way to go, but that isn't nice either since it's not very flexible with regard to adding/removing widgets

Reply:
> Do you think there is a way we could specify widgets within the format string? That seems a bit overkill

Translating formatting string like `"text {var} {var2}"` into four widgets seems overkill indeed. But maybe it is not that bad?

Reply:
I was thinking of something like `"{{ping}} {{up} {down}}"` or `"{ping} {|} {up} {down}"`. But i have absolutely no idea about the viability of this

Reply:
I mean... 714b29f1f0244c976780a0bf9c35c8fe9bbd6149 kinda works.

Reply:
It does, and quite well. Just some spacing to figure out:
![image](https://user-images.githubusercontent.com/9497832/111375144-378eeb00-869e-11eb-811e-e2f14a38cd3a.png)

Reply:
@MaxVerevkin on phone now so hard to check but will render_widgets still allow per widget click actions to work? (For when applying the same method to other blocks such as Nvidia)

Reply:
> will render_widgets still allow per widget click actions to work?

Yes, but I haven't tried yet. That's how I see it: if I wanted to handle a click on ping separately from the rest of the block, I can set it's `instance` right after I set it's status.

Reply:
I'll merge 714b29f then.